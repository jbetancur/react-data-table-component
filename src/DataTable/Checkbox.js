import React from 'react';
import PropTypes from 'prop-types';
import { handleFunctionProps } from './util';

const baseCheckboxStyle = { fontSize: '18px', cursor: 'pointer', marginLeft: '9px' };

const Checkbox = ({
  component,
  componentOptions,
  indeterminate,
  checked,
  name,
  onClick,
}) => {
  const setCheckboxRef = checkbox => {
    if (checkbox) {
      // eslint-disable-next-line no-param-reassign
      checkbox.indeterminate = indeterminate;
    }
  };

  const TagName = component;
  const baseStyle = TagName !== 'input' ? componentOptions.style : baseCheckboxStyle;
  const resolvedComponentOptions = handleFunctionProps(componentOptions, indeterminate);

  return (
    <TagName
      type="checkbox"
      {...resolvedComponentOptions}
      // allow this component to fully control these options
      ref={setCheckboxRef}
      style={baseStyle}
      onClick={onClick}
      name={name}
      aria-label={name}
      checked={checked}
      onChange={() => null} // prevent uncontrolled checkbox warnings -  we don't need onChange
    />
  );
};

Checkbox.propTypes = {
  indeterminate: PropTypes.bool,
  component: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  componentOptions: PropTypes.object,
  checked: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onClick: PropTypes.func,
};

Checkbox.defaultProps = {
  indeterminate: false,
  component: 'input',
  componentOptions: {
    style: {},
  },
  checked: false,
  onClick: null,
};

export default Checkbox;
