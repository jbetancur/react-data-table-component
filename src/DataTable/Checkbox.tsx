import * as React from 'react';
import { handleFunctionProps, noop } from './util';

const defaultComponentName = 'input';

const calculateBaseStyle = (disabled: boolean) => ({
	fontSize: '18px',
	...(!disabled && { cursor: 'pointer' }),
	padding: 0,
	marginTop: '1px',
	verticalAlign: 'middle',
	position: 'relative',
});

interface CheckboxProps {
	name: string;
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	component?: any;
	componentOptions?: { [key: string]: unknown };
	indeterminate?: boolean;
	checked?: boolean;
	disabled?: boolean;
	onClick?: (e: React.MouseEvent) => void;
}

function Checkbox({
	name,
	component = defaultComponentName,
	componentOptions = { style: {} },
	indeterminate = false,
	checked = false,
	disabled = false,
	onClick = noop,
}: CheckboxProps): JSX.Element {
	const setCheckboxRef = (checkbox: HTMLInputElement) => {
		if (checkbox) {
			// eslint-disable-next-line no-param-reassign
			checkbox.indeterminate = indeterminate;
		}
	};

	const TagName = component;
	const baseStyle = TagName !== defaultComponentName ? componentOptions.style : calculateBaseStyle(disabled);
	const resolvedComponentOptions = React.useMemo(
		() => handleFunctionProps(componentOptions, indeterminate),
		[componentOptions, indeterminate],
	);

	return (
		<TagName
			// allow this component to fully control these options
			type="checkbox"
			ref={setCheckboxRef}
			style={baseStyle}
			onClick={disabled ? noop : onClick}
			name={name}
			aria-label={name}
			checked={checked}
			disabled={disabled}
			{...resolvedComponentOptions}
			onChange={noop} // prevent uncontrolled checkbox warnings -  we don't need onChange
		/>
	);
}

export default React.memo(Checkbox);
