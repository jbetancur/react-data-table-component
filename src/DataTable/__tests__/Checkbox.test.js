import 'jest-styled-components';
import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import Checkbox from '../Checkbox';

afterEach(cleanup);

test('should render correctly when a custom component is not provided', () => {
  const { container } = render(<Checkbox name="test" />);

  expect(container.firstChild).toMatchSnapshot();
});

test('component <Checkbox component/> should render correctly with a custom checkbox', () => {
  // eslint-disable-next-line react/prefer-stateless-function
  class CustomComp extends React.Component {
    render() {
      return (<div>25 schmeckles</div>);
    }
  }

  const { container } = render(<Checkbox name="test" component={CustomComp} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('component <Checkbox component/> should render correctly with custom props', () => {
  const { container } = render(<Checkbox name="test" componentOptions={{ test: 'false' }} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('component <Checkbox indeterminate /> should toggle indeterminate to true on the element', () => {
  const { container } = render(<Checkbox name="test" indeterminate />);

  expect(container.firstChild.indeterminate).toBe(true);
});

test('component <Checkbox indeterminate={false} /> should not toggle indeterminate if there is no change', () => {
  const { container } = render(<Checkbox name="test" indeterminate={false} />);

  expect(container.firstChild.indeterminate).toBe(false);
});

test('should handle onClick', () => {
  const mockCallback = jest.fn();
  const { container } = render(<Checkbox name="test" onClick={mockCallback} />);
  fireEvent.click(container.firstChild);

  expect(mockCallback).toBeCalled();
});
