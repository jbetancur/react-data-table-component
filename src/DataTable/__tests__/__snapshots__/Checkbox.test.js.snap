// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component <Checkbox component/> should render correctly with a custom checkbox 1`] = `
<div>
  25 schmeckles
</div>
`;

exports[`component <Checkbox component/> should render correctly with custom props 1`] = `
<input
  aria-label="test"
  name="test"
  style="font-size: 18px; margin-left: 9px; cursor: pointer;"
  test="false"
  type="checkbox"
/>
`;

exports[`should render correctly when a custom component is not provided 1`] = `
<input
  aria-label="test"
  name="test"
  style="font-size: 18px; margin-left: 9px; cursor: pointer;"
  type="checkbox"
/>
`;
